libcob/profiling.c

```diff
@@ -0,0 +1,526 @@
+   as published by the Free Software Foundation, either version 3 of the
+   License, or (at your option) any later version.
+
+   GnuCOBOL is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with GnuCOBOL.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+
+#include "config.h"
+
+/* include internal and external libcob definitions, forcing exports */
+#define COB_LIB_EXPIMP
+#include "coblocal.h"
+
+#include "tarstamp.h"
+#include "common.h"
+
+#include <sys/types.h>
+#ifdef HAVE_UNISTD_H
+#include <unistd.h>
+#endif
+
+#ifdef	_WIN32
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
+#endif
+
+/* Local types and variables */
+
+struct cob_prof_module_list {
+	struct cob_prof_module *info ;
+	struct cob_prof_module_list *next;
+};
+
+static struct cob_prof_module_list *prof_info_list ;
+
+/* We maintain a stack of the procedures entered as 3 different
+ * arrays, with "current_idx" being the stack pointer. */
+static cob_ns_time              *start_times;
+static int                      *called_procedures;
+static struct cob_prof_module*  *called_runtimes;
+/* Current size of previous arrays */
+static int max_prof_depth;
+static int current_idx = -1;
+
+/* Whether profiling is active or not. */
+static int is_active = 0;
+/* Whether we are in testsuite mode */
+static int is_test = 0;
+
+/* Remember static and dynamic configuration */
+static cob_global               *cobglobptr = NULL;
+static cob_settings             *cobsetptr = NULL;
+
+
+
+/* Return the current time in nanoseconds. The result is guarranteed
+ * to be monotonic, by using an internal storage of the previous
+ * time. */
+static cob_ns_time
+get_ns_time (void)
+{
+	static cob_ns_time ns_time = 0;
+	if (is_test){
+		ns_time += 1000000;
+		return ns_time;
+	} else {
+		unsigned long long nanoseconds;
+#ifdef _WIN32
+		LARGE_INTEGER performance_counter;
+		LARGE_INTEGER performance_frequency;
+		QueryPerformanceCounter(&performance_counter);
+		QueryPerformanceFrequency(&performance_frequency);
+		performance_counter.QuadPart *= 1000000000;
+		performance_counter.QuadPart /= performance_frequency.QuadPart;
+		nanoseconds = performance_counter.QuadPart;
+		if (nanoseconds>ns_time) ns_time = nanoseconds;
+		return ns_time;
+#else
+#if defined (HAVE_CLOCK_GETTIME)
+		struct timespec ts;
+		clock_gettime(CLOCK_MONOTONIC, &ts);
+		nanoseconds = ts.tv_sec * 1000000000 + ts.tv_nsec;
+#else
+		nanoseconds = clock() * 1000000000 / CLOCKS_PER_SEC;
+#endif /* HAVE_CLOCK_GETTIME */
+#endif /* _WIN32 */
+		if (nanoseconds>ns_time) ns_time = nanoseconds;
+		return ns_time;
+	}
```

Comment:
This complete block has currently no coverage at all - we should have at least one test that disables the test mode (I _think_ that's possible with prog1, not compiled with profiling, doing a `SET ENVIRONMENT 'COB_IS_RUNNING_IN_TESTMODE' TO SPACE`, then calling progprof).